### PR TITLE
Remove cassandra-2x-compat and protocol-version config, #644

### PR DESCRIPTION
**Pull request by patriknw**
* instead use the ProtocolVersion from the driver

Refs #644

I'll run tests with Cassandra 2.x

### CODE REVIEW

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 1]
```diff
@@ -478,59 +483,67 @@ import akka.stream.scaladsl.Source
   private def delete(persistenceId: String, toSequenceNr: Long): Future[Unit] = {
 
     def physicalDelete(lowestPartition: Long, highestPartition: Long, toSeqNr: Long): Future[Done] = {
-      if (settings.cassandra2xCompat) {
-        def asyncDeleteMessages(partitionNr: Long, messageIds: Seq[MessageId]): Future[Unit] = {
-          val boundStatements = messageIds.map(mid =>
-            preparedDeleteMessages.map(_.bind(mid.persistenceId, partitionNr: JLong, mid.sequenceNr: JLong)))
-          Future.sequence(boundStatements).flatMap { stmts =>
-            executeBatch(batch => stmts.foldLeft(batch) { case (acc, next) => acc.add(next) })
+      session.protocolVersion.flatMap { version =>
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: most of it is whitespace indentation because of flatMap

**Comment by patriknw** [comment]
This didn't work out as intended. The `ProtocolVersion` from the driver is still v4 when using Cassandra server 2.2.15. That is as documented in https://docs.datastax.com/en/developer/java-driver/3.7/manual/native_protocol/

However, some features are not supported by 2.2.x. For example CassandraJournalSpec fails with 
```
InvalidQueryException: Invalid operator >= for PRIMARY KEY part sequence_nr,3
```

I guess that means we still need the  `cassandra-2x-compat` config flag?

**Comment by chbatey** [comment]
That's a shame. There's also CQL_VERSION doesn't give us the info we need

**Comment by chbatey** [comment]
It is in the table `system.local` so we could get it from there, we should be sure to deal with it not being there for the fake cassandras

**Comment by patriknw** [comment]
Thanks, I'll try to grab it from `system.local`.

There might be more specific features that we need to have more special cases for the fake cassandras, but I'm thinking we can add more specific support for them when we decide to support them, rather than treating them as 2.2.x.

**Comment by chbatey** [comment]
Yep default to not special casing them but handle the system.local table not existing

**Comment by patriknw** [comment]
Retrieving from system.local works fine. See latest commit.

**Comment by patriknw** [comment]
Tested with Cassandra 2.2.15 and many essential tests (where we have the special casing) works fine. Some failures, but tracking that in #691

File: session/src/main/scala/akka/stream/alpakka/cassandra/scaladsl/CassandraSession.scala [hunk 2]
```diff
@@ -88,16 +99,43 @@ final class CassandraSession(
...
+  /**
+   * Meta data about the Cassandra server, such as its version.
    */
-  def protocolVersion: ProtocolVersion =
-    underlying().value match {
-      case Some(Success(s)) =>
-        s.getContext.getProtocolVersion
-      case _ =>
-        throw new IllegalStateException("protocolVersion can only be accessed after successful init")
+  def serverMetaData: Future[CassandraServerMetaData] = {
```
Comment (chbatey) [inline session/src/main/scala/akka/stream/alpakka/cassandra/scaladsl/CassandraSession.scala]: 👍

File: session/src/main/scala/akka/stream/alpakka/cassandra/CassandraServerMetaData.scala [hunk 3]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.cassandra
+
+class CassandraServerMetaData(val clusterName: String, val dataCenter: String, val version: String) {
```
Comment (chbatey) [inline session/src/main/scala/akka/stream/alpakka/cassandra/CassandraServerMetaData.scala]: final

File: core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala [hunk 4]
```diff
@@ -206,35 +209,36 @@ import akka.stream.alpakka.cassandra.scaladsl.{ CassandraSession, CassandraSessi
...
-            val boundStatementBatches = mds
-              .map(md => preparedDeleteSnapshot.map(_.bind(md.persistenceId, md.sequenceNr: JLong)))
-              .grouped(0xFFFF - 1)
-            if (boundStatementBatches.nonEmpty) {
-              Future
-                .sequence(
-                  boundStatementBatches.map(
-                    boundStatements =>
+    session.serverMetaData.flatMap { meta =>
+      if (meta.isVersion2
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala]: 👍

File: session/src/main/scala/akka/stream/alpakka/cassandra/javadsl/CassandraSession.scala [hunk 5]
```diff
@@ -70,6 +73,18 @@ final class CassandraSession(@InternalApi private[akka] val delegate: scaladsl.C
    */
   def close(executor: Executor): CompletionStage[Done] = delegate.close(ExecutionContext.fromExecutor(executor)).toJava
 
+  /**
+   * The `ProtocolVersion` used by the driver to communicate with the server.
+   */
+  def protocolVersion: CompletionStage[ProtocolVersion] =
```
Comment (chbatey) [inline session/src/main/scala/akka/stream/alpakka/cassandra/javadsl/CassandraSession.scala]: do we need this now? can't hurt

Comment (patriknw) [reply session/src/main/scala/akka/stream/alpakka/cassandra/javadsl/CassandraSession.scala]: removed, available via underlying CqlSession if needed

**Pull request merged**